File: classes.dtree.r
# Set a min number of rows! Overfitting
# Level control
# Purity control
# random forest

data <- read.csv("data/classes.csv", header=TRUE)

cols <- dim(data)[2]
rows <- dim(data)[1]

train_data <- data[ 1:(0.15*rows), ]
test_data <- data[ (0.75*rows):rows, ]

# we are predicting the last attribute
# neg (0) to the left, pos (1) to the right
# node: list( left_node, right_node, index )

gen_node <- function(left_node, right_node, index){

  return( list(left_node, right_node, index) )

}

# purity: prob of 1
calc_purity <- function(cdata, index){

  pred_index <- dim(cdata)[2]

  ldata <- cdata[ cdata[,index]==0 ,]
  rdata <- cdata[ cdata[,index]==1 ,]

  lpos <- ldata[ ldata[,pred_index]==1 ,]
  ltotal <- dim(ldata)[1]
  lpos_total <- dim(lpos)[1]

  rpos <- rdata[ rdata[,pred_index]==1 ,]
  rtotal <- dim(rdata)[1]
  rpos_total <- dim(rpos)[1]

  return( (lpos_total/ltotal)^2 + (rpos_total/rtotal)^2 )

}

get_left_leaf <- function(cdata, index){
  ldata <- cdata[ cdata[,index]==0 ,]
  return(ldata)
}

get_right_leaf <- function(cdata, index){
  rdata <- cdata[ cdata[,index]==1 ,]
  return(rdata)
}

gen_tree <- function(cdata, level){

  print(level)

  crows <- dim(cdata)[1]
  ccols <- dim(cdata)[2]

  if(ccols==1 || crows==1){
    return(gen_node(-1,-1, -1))
  }

  purities <- unlist( lapply( c(1:ccols), function(x) calc_purity(cdata, x) ) )

  # we can use as root
  pure_node <- which.max( purities[ 1:(length(purities)-1) ] )

  left_cdata <- get_left_leaf(cdata, pure_node)[,-pure_node]
  level[1]<-level[1]+1
  left_node <- gen_tree(left_cdata, level)

  right_cdata <- get_right_leaf(cdata, pure_node)[,-pure_node]
  level[2]<-level[2]+1
  right_node <- gen_tree(right_cdata, level)

  return(gen_node(left_node, right_node, pure_node))

}

root <- gen_tree(train_data[,2:cols], c(1,1))

print(root)

File: classes.dtree.m
data = csvread("data/classes.csv");

% remove header and line count
data = data(2:end, 2:end);

cols_count = columns(data);
rows_count = rows(data);

train_data = data(1:(0.75*rows_count),:);
test_data = data((0.75*rows_count):end,:);

global max_level = 15;

function node = gen_node(left_node, right_node, index)

  node = struct(
    'left', left_node,
    'right', right_node
  );

end

function purity = calc_purity(cdata, index)

  pred_index = columns(cdata);

  ldata = cdata( cdata(:,index)==0 , : );
  rdata = cdata( cdata(:,index)==1 , : );

  lpos = ldata( ldata(:,pred_index)==1 , : );
  ltotal = rows(ldata);
  lpos_total = rows(lpos);

  rpos = rdata( rdata(:,pred_index)==1 , : );
  rtotal = rows(rdata);
  rpos_total = rows(rpos);

  purity = ( (lpos_total/ltotal)^2 + (rpos_total/rtotal)^2 );

end

function ldata = get_left_leaf(cdata, index)
  ldata = cdata( cdata(:,index)==0 , : );
end

function rdata = get_right_leaf(cdata, index)
  rdata = cdata( cdata(:,index)==1 , : );
end

function tree = gen_tree(cdata, level)

  % disp(level);

  global max_level;

  ccols = columns(cdata);
  crows = rows(cdata);

  if (ccols==1 || crows==1 || level(1,1)>=max_level || level(1,2)>max_level)
    tree = 0;
    return;
  end

  purities = rowfun( @(x) ( calc_purity(cdata, x) ) , [1:ccols]' );

  purities = purities(1:end-1,:);

  [plist,pindexes] = sort(purities, 'descend');

  pure_node = pindexes(1);

  left_cdata = get_left_leaf(cdata, pure_node);
  left_cdata(:,pure_node) = [];
  level(1) = level(1)+1;
  left_node = gen_tree(left_cdata, level);

  right_cdata = get_right_leaf(cdata, pure_node);
  right_cdata(:,pure_node) = [];
  level(2) = level(2)+1;
  right_node = gen_tree(right_cdata, level);

  tree = gen_node(left_node, right_node, pure_node);

end

root = gen_tree(train_data, [1 1]);

% root;
